lib/xenops_server.ml

```diff
@@ -2153,8 +2153,20 @@ and perform_exn ?subtask ?result (op : operation) (t : Xenops_task.task_handle)
             let final_handshake () =
               Handshake.send ~verbose:true mem_fd Handshake.Success ;
               debug "VM.migrate: Synchronisation point 3" ;
-              Handshake.recv_success mem_fd ;
-              debug "VM.migrate: Synchronisation point 4"
+              match Handshake.recv mem_fd with
+              | Success ->
+                  debug "VM.migrate: Synchronisation point 4"
+              | Error msg ->
+                  error
+                    "VM.migrate: Failed during Synchronisation point 4. msg: %s"
+                    msg ;
+                  ( try
+                      B.VM.remove vm ;
```

Comment:
B.VM.remove doesn't clean up the entire runtime state, I think we need the `VM_remove` `perform_atomics` operation for that (it also cleans up vbds, vifs, etc.), and in order to call that we need to hard shutdown the VM, see the exception handling code that you also modified.
We may also want to honour actions-after-crash, can we reuse VM_check_state here in some way?

Reply:
If we get a failure at this point, then the VM would have already had all its memory transferred and be in a suspended state. To clean it up properly, we would need similar actions at the end of the success case (below): `VM_shutdown` and `VM_remove`. The latter just makes xenopsd "forget" about the VM, while the former actually cleans up the VMs devices and domain.

I also thought that we would be able to rely on the auto-cleanup code in `VM_check_state`, but I'm not sure what the result of `get_domain_action_request` is in this state.

Reply:
I'm just trying these changes. I think the we reason `VM_check_state` isn't working is because it gets passed the original VM's uuid, rather than the new one that ends in 00001

Reply:
Looking at this part of the `trigger_cleanup_after_failure ` function: https://github.com/xapi-project/xenopsd/blob/master/lib/xenops_server.ml#L1915-L1920, it looks like you are right: the receive side distinguishes between the two uuids, but the send side (`VM_migrate`) doesn't. For `VM_migrate`, the `VM_check_state` is scheduled twice for the same uuid, which was correct in the past, but not anymore I think.

Reply:
The temporary uuid isn't part of the argument of `VM_migrate`. To achieve this, perhaps we need to derive the temporary uuids earlier, before scheduling the op in https://github.com/xapi-project/xenopsd/blob/96288f24aa66c6f3b4157dc208aadcf2d2073e5a/lib/xenops_server.ml#L3001-L3002